File: 0020 有效的括号/main.py
class Solution:
    def isValid(self, s: str) -> bool:
        a=[]
        for i in s:
            if '['==i or '('==i or "{"==i:
                a.append(i)
            else:
                if len(a)==0:
                    return False
            if '}'==i:
                if a.pop() !="{":
                    return False
            
            if ')'==i:
                if a.pop() !="(":
                    return False
            
            if ']'==i:
                if a.pop() !="[":
                    return False
        return True and len(a)==0

File: 0020 有效的括号/main.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_set<char> leftPool={'{','(','['};
        unordered_map<char,char> rightPool={{'}','{'},{')','('},{']','['}};
        string stk;
        for(int i =0;i<s.size();i++){
            if( leftPool.find(s[i]) != leftPool.end() ){
                stk.push_back(s[i]);
            }else{
                if( stk.size() >0 && stk.back() == rightPool[ s[i] ]){
                    stk.pop_back();
                }else{
                    return false;
                }
            }
        }
        return stk.size()==0;
    }
};